### ISSUE TITLE
Support some POSIX-RT APIs to run rt-tests (cyclictest) on Osv

### ISSUE THREAD

**Issue by hw-claudio** [cloudius-systems/osv#386]
I have tried building and running cyclictest on OSv:

https://rt.wiki.kernel.org/index.php/Cyclictest

git clone git://git.kernel.org/pub/scm/linux/kernel/git/clrkwllms/rt-tests.git
or tarballs:

https://www.kernel.org/pub/linux/kernel/people/clrkwllms/

there seem to be these missing functions required to be able to do a minimal cyclictest run:

sched_getscheduler
sched_setscheduler

timer_create
timer_settime

do you think it's possible to implement these POSIX rt functions?

Ciao,

Claudio

**Comment by bamboojsr** [comment]
@nyh  @hw-claudio , Today I met same problem, But I had add pthread into OSv lib. 

Adding /libenviron.so...
Adding /zpool.so...
Adding /libzfs.so...
Adding /libuutil.so...
Adding /zfs.so...
Adding /tools/mkfs.so...
Adding /tools/cpiod.so...
Adding /tools/mount-nfs.so...
Adding /tools/umount.so...
Adding /usr/lib/libgcc_s.so.1...
Adding /usr/lib/libpthread.so.0...
Adding /usr/lib/libboost_filesystem.so.1.55.0...
Adding /usr/lib/libboost_regex.so.1.55.0...
Adding /etc/hosts...
Link /etc/mnttab to /proc/mounts ...
Adding /etc/fstab...
Adding /dev...
Adding /proc...
Adding /tmp...
Adding /tools/cyclictest.so...

**Comment by nyh** [comment]
libpthread.so.0 is on the list of objects supplied by the OSv kernel itself (see the full list in program::program in elf.cc), so adding it to the image should not make a difference (did it?), and even if it did, it is quite unlikely that anything from libpthread.so (which is part of glibc) will actually work unmodified on OSv.

If you're interested in adding real-time scheduling to OSv (can you please say something about your use case?) you might be happy to know that we actually have working patches for many (but not all) of these features: I have a patch (which I continue to maintain off-tree) adding OSv-specific methods set_realtime_priority(), and set_realtime_slice(), and Claudio had patches to use those to implement some of the Posix API for scheduling classes.

My patch involves tinkering inside the OSv scheduler - it is not something as simple as strapping on an additional library like libthread.so.

**Comment by bamboojsr** [comment]
Yes, Wireless workload 's scheduling in 1ms, So  real-time is very important, we don't want to waste more time at OS schedule.  In current time ,we find guest's real-time more than native os(aboute 3doubles). 
So I want to know the  OSv real-time performance by cyclictest .

**Comment by hw-claudio** [comment]
Hi all,

I will look at the state of the current RT scheduler patchset from Nadav and my RT API branch when I get back next week.

I got cyclictest working at the time using them.

Ciao!

Claudio

**Comment by nyh** [comment]
On Tue, Feb 9, 2016 at 5:58 AM, Claudio Fontana <notifications@github.com>
wrote:

> Hi all,
>
> I will look at the state of the current RT scheduler patchset from Nadav
> and my RT API branch when I get back next week.
>

> I got cyclictest working at the time using them.
>

Thanks Claudio. I'll post my part of the patch again (rebased to the latest
code) in a few days.

**Comment by nyh** [comment]
sched_getscheduler() is needed for Python 3.6, which is apparently (?) compiled with "-z now" so needs a lot of symbols to be available, but doesn't quite care what they do. @wkozaczuk

**Comment by nyh** [comment]
We should start by trivially implementing sched_getscheduler() and friends for the normal, already implemented, SCHED_OTHER. Only later we can worry about the patches I sent to the mailing list, for supporting SCHED_FIFO and incomplete support for SCHED_RR.

**Comment by nyh** [comment]
If we want to revive the realtime scheduler feature in OSv, here is my old scheduler patch:
https://groups.google.com/d/msg/osv-dev/cdf5jYSsCuc/bENVrfuX6EkJ
this patch adds a new API for controlling the realtime scheduler, and here is @hw-claudio 's patches for implementing the relevant Posix API (sched_getscheduler() et al.): 
https://groups.google.com/d/msg/osv-dev/3QN0uZkW9jA/685VZQS9WFYJ

**Comment by nyh** [comment]
I rebased by low-level scheduler patch as a pull-request, here: https://github.com/cloudius-systems/osv/pull/1223/

We also need @hw-claudio's patches (see above links, or last few patches in https://github.com/hw-claudio/osv_aarch64/commits/rt_wip https://github.com/hw-claudio/osv_aarch64/commits/rt) to provide the POSIX API compatibility on top of my low-level support.

**Comment by lucabe72** [comment]
Thanks!
I am going to test this patch (together with Claudio's patches) in the next few days.

Luca

**Comment by lucabe72** [comment]
Sorry about the delay... I updated my build machine and had some trouble compiling OSv.
Anyway, now that I fixed these issues, I've been able to test this patch and to forward-port Claudio's patches. I needed to change them a little bit, but I can finally start cyclictest!!!
The cyclictest command line I use is
    cyclictest -i 1000 -n -q

The measured latencies are not good, but this is another story :)


Thanks,
Luca

**Comment by lucabe72** [comment]
I forgot: after Easter, I'll post the updated patches I am using

**Comment by nyh** [comment]
> The measured latencies are not good, but this is another story :)

After we get the patches in, let's discuss these "measured latencies". I'm not familiar with cyclictest, what latency does it measure? How "bad" is it?
Note that if it relies on SCHED_RR (on Linux the default, /proc/sys/kernel/sched_rr_timeslice_ms, is 100ms time slices) then we don't support that yet.

**Comment by lucabe72** [comment]
On Sun, 9 Apr 2023 at 17:11, nyh ***@***.***> wrote:

> The measured latencies are not good, but this is another story :)
>
> After we get the patches in, let's discuss these "measured latencies".
>
I agree, let's discuss this after I post the updated patches (and they are
committed if they are OK).
Also, maybe this is not the best place to discuss latency issues (maybe I
should start a new thread on a mailing list?)

I'm not familiar with cyclictest, what latency does it measure? How "bad"
> is it?
>
Cyclictest just starts one (or more) periodic thread with a high priority
(so that the thread is supposed to be scheduled immediately at every
periodic wakeup), and measures the difference between the theoretical
wakeup time and the time when the thread is scheduled.
From my first non-systematic tests, this difference was various
milliseconds (this is not surprising, because if I understand well this
kind of latency has never been measured and optimized before). But I need
to run some better tests before having numbers that I can discuss with
confidence.

Note that if it relies on SCHED_RR (on Linux the default,
> /proc/sys/kernel/sched_rr_timeslice_ms, is 100ms time slices) then we don't
> support that yet.
>
Yes, I noticed that... But AFAIK cyclictest just needs SCHED_FIFO


Thanks,
Luca

**Comment by nyh** [comment]
> Also, maybe this is not the best place to discuss latency issues (maybe I should start a new thread on a mailing list?)

Right.

> I'm not familiar with cyclictest, what latency does it measure? How "bad" is it?
> Cyclictest just starts one (or more) periodic thread with a high priority (so that the thread is supposed to be scheduled immediately at every periodic wakeup), and measures the difference between the theoretical wakeup time and the time when the thread is scheduled. From my first non-systematic tests, this difference was various milliseconds (this is not surprising, because if I understand well this kind of latency has never been measured and optimized before).

Please also try to run the same test on a single CPU (if you use scripts/run.py, use "run.py -c1" for one CPU, "run.py -c2" for two CPUs, etc.) and see if you get different results. And let's, indeed, continue discussing this on the mailing list.

**Comment by lucabe72** [comment]
Hi,

I wanted to post here the patches I am using. But I see that I cannot attach patches as text files... I'll send an email to osv-dev@googlegroups.com with the patches.

**Comment by nyh** [comment]
Yes, you can send patches to the mailing list. Alternatively, you can create a pull request as I did in https://github.com/cloudius-systems/osv/pull/1223
